README.md
# Spotify Labelling
## Introduction
In my personal opinion, Spotify is a great platform for streaming music. However, it has one big downside: you cannot create "smart playlists" which create playlists based on one or multiple conditions. But that is no longer a limitation as that is exactly what Spotify Labelling offers us!



## Status
At this point, the project is still in the alpha phase (v0.14.0-alpha), where a few people are invited to test the application. The goal is to have a working application that can be released as beta in August or September. You can follow the progress by viewing the [changelog](https://github.com/BrentMeeusen/Spotify-Labelling/blob/master/CHANGELOG.md).



## The application
The goal of the project was to make it as intuitive as possible to use, but if you happen to be unsure how you should use the application, you can read the explanation below. If you have an account, you can also [log in](http://spotify-labelling.21webb.nl/) and visit [this page](http://spotify-labelling.21webb.nl/how-it-works).

### Create an account
To get started, you create an account [here](http://spotify-labelling.21webb.nl/register). After verifying your account by clicking on the link in the email you will receive, you can log in. When logging in, you have to grant the application access through Spotify. If you don't do that, the application cannot make any changes and will not let you login.

### Import your songs
Once you have logged in, you will start off by selecting the songs you want to add to Spotify Labelling. This can be done by clicking the "Add Songs" button which will then allow you to choose the playlists to import the songs from, or you can choose to import the songs you saved. You will now be able to view all these songs on the page where you can search for them by label, song name, artist name, and rating, as well as update the rating.

### Sort your songs
Now you can start creating your own labels. There are already some global labels as well that many songs may belong to, like "Rock" and "EDM". With your songs imported and labels created, you can start making combinations. You can now give a song a label which will be used to create the playlists.

### Create smart playlists
At this point, you have imported your songs and given them a rating, created labels, and added those labels to your songs. Now there is only one more thing left to do: creating the playlists! When creating a playlist, you will give it a name the playlist will have (which will be the name in this application as well as in your own Spotify) and then add the requirements the songs must meet before making it in the playlist. The last step is to click the "Update Playlists" button, which will send all the changes to Spotify.

### Updating your songs
It is of course very possible that you gave a song a label, but you do not find it fitting anymore, or you like the song less so you want to lower its rating. All you have to do then is make your changes and click the "Update Playlists" button again. This will update the already existing playlists to match the new values!



<!-- ## Test the software -->



## Known issues
At this point, I am not aware of any issues in v0.14.0-alpha.
<!-- In v0.14.0-alpha, I am aware of the following issues: -->



## Getting help
If you encounter problems with using the program, please create an issue with the tag `help`.



## Credits
This project has been created by me, but I have been helped by a couple of my friends who I asked to be alpha tester. My special thanks go to [Teambeurnaut](https://github.com/Teambeurnaut) for being the most dedicated alpha tester. He helped me find many bugs and came up with some small features which, in total, made the application better usable.




CHANGELOG.md
# Changelog
In here, I write the changelog for the project.

<!-- 

STANDARD NEW VERSION:

## x.y.z - YYYY-MM-DD
### Added
- a
- b
- c

### Changed
- a
- b
- c

### Removed
- a
- b
- c

### Fixed
- a
- b
- c

-->

<!-- ## 0.14.1-alpha - 2021/10/??

### Added
- Filtering your tracks on:
	- Artist is not
	- Does not have label

### Fixed
- Seeing the results of input sanitizing
 -->



## 0.14.0-alpha - 2021/09/18

### Added
- Reset password
- A field in the navigation menu where it displays with which Spotify account you're connected
- "Click/tap to dismiss" message to notification popup
- Showing how many songs have the label on the "Your Labels" page
- Creating, updating, getting, and deleting playlists
- Removing a label from all visible songs

### Changed
- All data that is not connected to any other user than the user that deletes its account, will no longer be stored
- The scopes that are required for the application to work

### Fixed
- Numbers in "Your songs" wouldn't reset after modifying labels



## 0.13.2-alpha - 2021/09/10

### Added
- Numbers in "Your songs" list so it's easier to remember where you left off

### Fixed
- After an invalid JSON request, you'd have to refresh the page to use the application again
- Trying to import local files (which is not possible)



## 0.13.1-alpha - 2021/08/30

### Added
- Importing your Liked Songs
- Alphabetical order for labels and songs

### Changed
- 404 to 200 status code when no labels are found

### Fixed
- Unkown artists bug when removing song from Your Songs
- Bug where label creator was null



## 0.13.0-alpha - 2021/08/24

### Added
- Filtering your tracks on:
	- Artist name
	- Track name
	- Label name
	- Number of labels
	- Released before/after
	- Added before/after
- Add labels to tracks (multiple labels at once)
- Remove labels from tracks
- Add labels to all selected tracks at once (multiple labels at once)

### Changed
- Closing popup on removing track from my list
- The way tracks are retrieved from the database



## 0.12.0-alpha - 2021/08/17

### Added
- Limited number of requests a user can make within one minute
- Being able to remove tracks from the application, including unnecessary data that is left over afterwards

### Changed
- Getting the tracks is now done in a different class
- Labels display
- Tracks display
- Importing playlists display
- The way routing is done



## 0.11.1-alpha - 2021/08/08

### Changed
- Preflight check due to CORS policies



## 0.11.0-alpha - 2021/08/08

### Added
- Classes for data from Spotify
- Class for Spotify data collection
- Classes for data from my own database
- Result popup disappears on click

### Changed
- Table methods into Database
- Label class
- User class

### Removed
- Table class
- Classes that worked with both Spotify and my data simultaneously



## 0.10.3-alpha - 2021/06/30

### Added
- Playlist size limit because the server can't handle over 2000 tracks
- Limit to label text length

### Fixed
- Caught Spotify API response errors
- A minor bug that has to do with query parameters
- Buttons no longer send requests until the last request has received a response



## 0.10.2-alpha - 2021/06/28

### Fixed
- Showing labels would stop after public label is not owned
- Authorisation tokens wouldn't be stored completely
- Started waiting for Spotify API to prevent 419 and 428 errors



## 0.10.1-alpha - 2021/06/21

### Added
- State token for Spotify to make it more secure

### Fixed
- Refactored Your Songs page
- Refactored CSS files
- Moved AddedAt timestamp column from Tracks to Tracks_to_users



## 0.10.0-alpha - 2021/06/19

### Added
- Creating a new payload from the database values as soon as the token is required so we are up-to-date
- Made user grant access to the application on login
- Load in user playlists
- Import all songs from selected playlist into the database
- Redirecting to login screen when the token expires and a request is made
- Landing page

### Changed
- Refactored tables and table creation
- Removed and re-added all tables, with it removing all existing data

### Fixed
- Conditions for TRUE being too strict (as 1 may also work in some specific situations)
- Logout including a file that was not found bug



## 0.9.1-alpha - 2021/06/11

### Fixed
- Verifying method that did not work with the new interface yet
- Errors on verifying will now be returned to the user



## 0.9.0-alpha - 2021/06/11

### Added
- Showing labels on the page
- Edit labels
- Delete labels

### Changed
- Allowed deleting accounts when labels are created

### Fixed
- Logging in, even when account is not verified yet



## 0.8.0-alpha - 2021/06/09

### Added
- Your labels layout
- Create label
- Can add labels through the user interface



## 0.7.1-alpha - 2021/05/12

### Changed
- Clicking on title now redirects to dashboard instead of the current page



## 0.7.0-alpha - 2021/05/12

### Added
- Your songs page
- Your labels page
- Your playlists page
- How it works page
- Navigation menu
- Autofills form when new data is received if it's set
- Logout button

### Changed
- Additional password rule on password update
- Additional input check on account update
- Create a new JWT on updating account



## 0.6.1-alpha - 2021/05/10

### Added
- Waits for response before the next request can be made

### Changed
- Registering password login rules
- Must insert account password in order to delete account



## 0.6.0-alpha - 2021/05/10

### Added
- Delete your account
- Logout mechanism



## 0.5.1-alpha - 2021/05/09

### Changed
- Return messages from the database



## 0.5.0-alpha - 2021/05/09

### Added
- values.js
- Your Account page
- Update your account

### Fixed
- Refactored register.js 



## 0.4.2-alpha - 2021/05/08

### Fixed
- Made JWT only accessible in cookie instead of the session too



## 0.4.1-alpha - 2021/05/08

### Fixed
- Page protection where payload.user was not set
- JSON Web Token set in JavaScript from the cookie instead of the session




AAA_API/routing.php
<?php

// Headers here because this file is added to all endpoints
header("Access-Control-Allow-Origin: http://spotify-labelling.21webb.nl");
header("Access-Control-Allow-Methods: GET, POST, PATCH, PUT, DELETE, OPTIONS");
header("Access-Control-Allow-Headers: Origin, Content-Type, X-Auth-Token, Authorization");
header("Content-Type: application/json; charset=UTF-8");

// Collect data
$url = $_SERVER["REQUEST_URI"];
$post = json_decode(file_get_contents("php://input"));
$jwt = (isset(getallheaders()["Authorization"]) ? @explode("Bearer ", getallheaders()["Authorization"])[1] : "");

// Include classes
include_once("classes/api-response.php");
include_once("classes/database.php");
include_once("classes/initialise.php");
include_once("classes/jwt.php");
include_once("classes/request.php");
include_once("classes/spotify-api.php");

// If it's a preflight check, return 200
if($_SERVER["REQUEST_METHOD"] === "OPTIONS") {
	ApiResponse::httpResponse(200);
}

// Include Spotify models
include_once("classes/models/spotify/spotify-collection.php");
include_once("classes/models/spotify/spotify-album.php");
include_once("classes/models/spotify/spotify-artist.php");
include_once("classes/models/spotify/spotify-playlist.php");
include_once("classes/models/spotify/spotify-track.php");

// Include my Spotify data models
include_once("classes/models/my/album.php");
include_once("classes/models/my/artist.php");
include_once("classes/models/my/playlist.php");
include_once("classes/models/my/track.php");

// Include general
include_once("classes/models/label.php");
include_once("classes/models/user.php");
include_once("classes/methods.php");

// Connect with the database
Database::connect();










/*************************
 ******** ROUTING ********
 ************************/

// If the URL does not contain `/api/`, return an error
if(strpos($url, "/api/") === FALSE) {
	ApiResponse::httpResponse(404, ["error" => "Page not found."]);
}

// Set base at /api/
$routes = explode("/", explode("/api/", $url)[1]);



// /api/create-tables
if(isset($routes[0]) && $routes[0] === "create-tables") {

	Request::checkRequestMethod(["POST"]);

	include_once("AAA_DEV/create-tables.php");

}	// /api/v1/create-tables



// /api/v1
if(isset($routes[0]) && $routes[0] === "v1") {

	// /api/v1/login
	if(isset($routes[1]) && $routes[1] === "login") {

		Request::checkRequestMethod(["POST"]);

		$identifier = @$post->Identifier;
		$password = @$post->Password;
		include_once("users/login.php");

	}	// /api/v1/login



	// /api/v1/register
	if(isset($routes[1]) && $routes[1] === "register") {

		Request::checkRequestMethod(["POST"]);
		include_once("users/register.php");

	}	// /api/v1/register



	// /api/v1/labels
	if(isset($routes[1]) && $routes[1] === "labels") {

		// /api/v1/labels/create
		if(isset($routes[2]) && $routes[2] === "create") {

			Request::checkRequestMethod(["POST"]);
			$payload = JSONWebToken::getPayload($jwt);

			$name = @$post->Name;
			$values = ["Name" => $name];
			include_once("labels/create.php");

		}	// /api/v1/labels/create

		// /api/v1/labels/delete
		if(isset($routes[3]) && $routes[3] === "delete") {

			Request::checkRequestMethod(["DELETE"]);
			$payload = Request::requireToken($jwt);

			$labelID = $routes[2];
			include_once("labels/delete.php");

		}	// /api/v1/labels/delete

		// /api/v1/labels/[label-id]/update
		if((isset($routes[3]) && $routes[3] === "update")) {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			$labelID = $routes[2];
			$name = @$post->Name;
			$values = ["Name" => $name];
			include_once("labels/update.php");

		}	// /api/v1/labels/[label-id]/update

		// /api/v1/labels
		Request::checkRequestMethod(["GET"]);
		$payload = Request::requireToken($jwt);

		$userID = @$routes[2];
		include_once("labels/get.php");

	}	// /api/v1/labels



	// /api/v1/playlists
	if(isset($routes[1]) && $routes[1] === "playlists") {

		// /api/v1/playlists/create
		if(isset($routes[2]) && $routes[2] === "create") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			$name = @$post->Name;
			include_once("playlists/create.php");

		}

		// /api/v1/playlists/[playlist-id]/delete
		if(isset($routes[3]) && $routes[3] === "delete") {

			Request::checkRequestMethod(["DELETE"]);
			$payload = Request::requireToken($jwt);

			$id = (isset($routes[3]) ? $routes[2] : NULL);
			include_once("playlists/delete.php");

		}	// /api/v1/playlists/delete

		// /api/v1/playlists/[playlist-id]/update
		if(isset($routes[3]) && $routes[3] === "update") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			$id = (isset($routes[3]) ? $routes[2] : NULL);
			$name = @$post->Name;
			$values = ["Name" => $name];
			include_once("playlists/update.php");

		}	// /api/v1/playlists/update

		// /api/v1/playlists
		Request::checkRequestMethod(["GET"]);
		$payload = Request::requireToken($jwt);

		$userID = @$payload->user->id;
		include_once("playlists/get.php");

	}	// /api/v1/playlists



	// /api/v1/spotify
	if(isset($routes[1]) && $routes[1] === "spotify") {

		// /api/v1/spotify/import/[playlist-id]
		if(isset($routes[2]) && $routes[2] === "import") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			$playlistID = @$routes[3];
			include_once("spotify/import.php");

		}	// /api/v1/spotify/import/[playlist-id]

		// /api/v1/spotify/playlists
		if(isset($routes[2]) && $routes[2] === "playlists") {

			Request::checkRequestMethod(["GET"]);
			$payload = Request::requireToken($jwt);

			include_once("spotify/playlists.php");

		}	// /api/v1/spotify/playlists

	}	// /api/v1/spotify



	// /api/v1/tracks
	if(isset($routes[1]) && $routes[1] === "tracks") {

		// /api/v1/tracks/add-labels/[track-id]
		if(isset($routes[2]) && $routes[2] === "add-labels") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);
			$trackID = @$routes[3];
			include_once("tracks/add-labels.php");

		}	// /api/v1/tracks/add-labels/[track-id]

		// /api/v1/tracks/remove-labels/[track-id]/[label-id]
		if(isset($routes[2]) && $routes[2] === "remove-labels") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);
			$trackID = @$routes[3];
			$labelID = @$routes[4];
			@$post->labelOne = $labelID;
			include_once("tracks/remove-labels.php");

		}	// /api/v1/tracks/remove-labels/[track-id]/[label-id]

		// /api/v1/tracks/get
		if(isset($routes[2]) && $routes[2] === "get") {

			Request::checkRequestMethod(["GET"]);
			$payload = Request::requireToken($jwt);

			// /api/v1/tracks/get/[track-id]
			if(isset($routes[3]) && $routes[3] !== "") {
				$trackID = @$routes[3];
			}
			include_once("tracks/get.php");

		}	// /api/v1/tracks/get

		// /api/v1/tracks/[track-id]/delete
		if(isset($routes[2]) && preg_match("/[a-zA-Z0-9]+/", $routes[2]) !== FALSE && isset($routes[3]) && $routes[3] === "delete") {

			Request::checkRequestMethod(["DELETE"]);
			$payload = Request::requireToken($jwt);

			$trackID = @$routes[2];
			include_once("tracks/delete.php");

		}	// /api/v1/tracks/[track-id]/delete

	}	// /api/v1/tracks



	// /api/v1/users
	if(isset($routes[1]) && $routes[1] === "users") {

		// /api/v1/users/add-token
		if(isset($routes[2]) && $routes[2] === "add-token") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			$token = @$post->AccessToken;
			include_once("users/add-token.php");

		}	// /api/v1/users/add-token

		// /api/v1/users/forgot-password
		if(isset($routes[2]) && $routes[2] === "forgot-password") {

			Request::checkRequestMethod(["POST"]);

			$emailAddress = @$post->{"email-address"};
			include_once("users/forgot-password.php");

		}	// /api/v1/users/forgot-password

		// /api/v1/users/set-password
		if(isset($routes[2]) && $routes[2] === "set-password") {

			Request::checkRequestMethod(["POST"]);

			$emailAddress = @$post->EmailAddress;
			$password = $post->Password;
			$userID = $post->UserID;
			include_once("users/set-password.php");

		}	// /api/v1/users/set-password

		// /api/v1/users/set-spotify-email
		if(isset($routes[2]) && $routes[2] === "set-spotify-email") {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			include_once("users/set-spotify-email.php");

		}	// /api/v1/users/set-spotify-email

		// /api/v1/users/create
		if(isset($routes[2]) && $routes[2] === "create") {

			Request::checkRequestMethod(["POST"]);
			$payload = JSONWebToken::getPayload($jwt);

			$email = @$post->EmailAddress;
			$password = @$post->Password;
			include_once("users/create.php");

		}	// /api/v1/users/create

		// /api/v1/users/delete
		if((isset($routes[2]) && $routes[2] === "delete") || (isset($routes[3]) && $routes[3] === "delete")) {

			Request::checkRequestMethod(["DELETE"]);
			$payload = Request::requireToken($jwt);

			$id = (isset($routes[3]) ? $routes[2] : NULL);
			$password = @$post->Password;
			include_once("users/delete.php");

		}	// /api/v1/users/delete

		// /api/v1/users/get
		if(isset($routes[2]) && $routes[2] === "get") {

			Request::checkRequestMethod(["GET"]);
			$payload = Request::requireToken($jwt);

			include_once("users/get.php");

		}	// /api/v1/users/get

		// /api/v1/users/update
		if((isset($routes[2]) && $routes[2] === "update") || (isset($routes[3]) && $routes[3] === "update")) {

			Request::checkRequestMethod(["POST"]);
			$payload = Request::requireToken($jwt);

			$id = (isset($routes[3]) ? $routes[2] : NULL);
			$email = @$post->EmailAddress;
			$password = @$post->Password;
			$values = ["EmailAddress" => $email, "Password" => $password];
			include_once("users/update.php");

		}	// /api/v1/users/update

		// /api/v1/users/verify
		if(isset($routes[2]) && $routes[2] === "verify") {

			Request::checkRequestMethod(["GET"]);

			$id = @$routes[3];
			$email = @$routes[4];
			include_once("users/verify.php");

		}	// /api/v1/users/verify

	}	// /api/v1/users

}	// /api/v1

// If there's no match, return an error
ApiResponse::httpResponse(404, ["error" => "Page not found."]);

// exit();

?>